x-pack/platform/plugins/shared/streams/server/lib/streams/state_management/stream_active_record.ts

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { IScopedClusterClient, Logger } from '@kbn/core/server';
+import { StreamDefinition } from '@kbn/streams-schema';
+import { State, StreamChange } from './state'; // Does this create a circular dependency?
+import { StreamsStorageClient } from '../service';
+import { AssetClient } from '../assets/asset_client';
+import { ElasticsearchAction } from './execution_plan';
+import { StreamsClient } from '../client';
+
+export interface StreamDependencies {
+  scopedClusterClient: IScopedClusterClient;
+  assetClient: AssetClient;
+  storageClient: StreamsStorageClient;
+  logger: Logger;
+  isServerless: boolean;
+  streamsClient: StreamsClient;
+}
+
+export interface ValidationResult {
+  isValid: boolean;
+  errors: string[]; // Or Errors?
+}
+
+type StreamChangeStatus = 'unchanged' | 'upserted' | 'deleted';
+
+export abstract class StreamActiveRecord<TDefinition extends StreamDefinition = StreamDefinition> {
+  protected _updated_definition: TDefinition;
+  protected dependencies: StreamDependencies;
+  protected changeStatus: StreamChangeStatus = 'unchanged';
+
+  constructor(definition: TDefinition, dependencies: StreamDependencies) {
+    this._updated_definition = definition;
+    this.dependencies = dependencies;
+  }
+
+  public get definition(): TDefinition {
+    return this._updated_definition;
+  }
+
+  async applyChange(
```

Comment:
We should document somehow that the stream should focus on itself and it's direct parent/children